reader-privacy.md

```diff
@@ -0,0 +1,199 @@
+over the decrypted `peerID` part of it. Using such hash for each `ProviderRecordKey` the client would do another lookup 
+to get an encrypted `ProviderRecord` in response. `ProviderRecord` will contain information about the provider, 
+such as it's *peerID*, *multiaddrs*, *supported protocols* and so on. Each `ProviderRecord` will be encrypted 
+with a key derived from `peerID`. In order to make sense of that payload, a passive observer would need to 
+get hold of the original `peerID` that isn't revealed during the communication round;
+* Using a key derived from `peerID`, the client will decrypt `ProviderRecord`s and then reach out to the 
+provider directly to fetch the desired content. 
+
+By utilising such scheme only a party that knows original CID can decode the protocol,
+and that CID is never revealed. 
+
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant indexer
+    participant provider
+    client->>client: calculates hash(MH)
+    client->>indexer: sends a find request for hash(MH)
+    indexer->>client: sends a list of [ProviderRecordKey], each encrypted with a key derived from MH
+    loop ProviderRecordKeys
+        client->>client: decrypts ProviderRecordKey and extracts peerID from it
+        client->>indexer: sends ProviderRecord lookup request for hash(peerID)
+        indexer->>client: sends a ProviderRecord encrypted with a key derived from peerID
+        client->>client: decrypts the ProviderRecord
+    end
+    client->>provider: reaches out to the provider for the desired content
+```
+
+
+### Security
+​
+Security model of the Reader Privacy proposal boils down to inability to *algorithmically* derive the original multihash value for a 
+`hash(multihash)` that is used for IPNI lookups. Right now advertisments are not encrypted, but authenticated and contain plain multihash values in them. 
+That is going to change once *Writer Privacy* is implemented. Until then, an attacker could build a map of `hash(multihash) -> multihash` 
+by re-ingesting advertisements chain from each publisher in order to collect all original multihashes which can then be used to decrypt provider records and so on. 
+Doing that will require significant resources as it involves crawling the entire network. However, it will eventually be eliminated by *Writer Privacy* upgrade.
+
+Reader Privacy is a first step towards fully private content routing protocol. 
+
+Wider security implications are discussed in the IPFS Reader Privacy specification: TODO link here.
+
+#### Hashing and Encryption Function Upgrades
+
+All multihashes have a codec encoded in them. If a hashing or encryption funciton will have to rotate then different types of multihahses can coexist together
+and can be processed differently by IPNI implementations. It won't be possible to apply a fix retroactivelly to the data returned by previous lookup requests, 
+however IPNI implementations should start blocking all new ones that use a compromised scheme.
```

Comment:
so IPNI will not allow retro compatibility with older client versions as soon as a fix is issued? Clients will need to upgrade to keep using the IPNI?

Reply:
Great spot! There should be some transition period of course. Rephrased to make it clearer.